File: examples/canonical-gcp-functions/gcs-trigger-ocr-to-firestore/python/src/handler.py
from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass
from typing import Any, Protocol


class FirestoreClient(Protocol):
    def claim_generation(self, dedupe_key: str, payload: dict[str, Any]) -> bool: ...
    def write_metadata(self, dedupe_key: str, payload: dict[str, Any]) -> None: ...


class VisionClient(Protocol):
    def ocr(self, bucket: str, name: str, generation: str) -> str: ...


class PubSubClient(Protocol):
    def publish(self, topic: str, payload: dict[str, Any]) -> None: ...


@dataclass
class Clients:
    firestore: FirestoreClient
    vision: VisionClient
    pubsub: PubSubClient


def handle_gcs_object(event: dict[str, Any], clients: Clients | None = None) -> dict[str, Any]:
    started = time.time()
    data = event.get("data", {})
    bucket = data["bucket"]
    name = data["name"]
    generation = str(data.get("generation") or data.get("metageneration") or "0")
    event_id = event.get("id", f"{bucket}/{name}/{generation}")
    dedupe_key = f"{bucket}:{name}:{generation}"
    clients = clients or _gcp_clients()

    claimed = clients.firestore.claim_generation(dedupe_key, {"bucket": bucket, "name": name, "generation": generation})
    if not claimed:
        _log(started, event_id, dedupe_key, "DROP", "duplicate object generation")
        return {"ok": True, "duplicate": True, "decision": "DROP"}

    text = clients.vision.ocr(bucket, name, generation)
    payload = {"bucket": bucket, "name": name, "generation": generation, "text": text, "status": "OCR_COMPLETE"}
    clients.firestore.write_metadata(dedupe_key, payload)
    clients.pubsub.publish(os.getenv("DOWNSTREAM_TOPIC", "accb-dev-gcp-ocr-downstream"), payload)
    _log(started, event_id, dedupe_key, "ALLOW", "ocr metadata written")
    return {"ok": True, "duplicate": False, "decision": "ALLOW", "text": text}


def _gcp_clients() -> Clients:
    raise RuntimeError("GCP clients are bound by deployment; tests inject clients.")


def _log(started: float, event_id: str, dedupe_key: str, decision: str, msg: str) -> None:
    print(
        json.dumps(
            {
                "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "severity": "INFO",
                "msg": msg,
                "trace_id": os.getenv("TRACEPARENT", "local-trace"),
                "request_id": event_id,
                "correlation_id": event_id,
                "provider": "gcp",
                "runtime_tier": "function",
                "function_name": os.getenv("FUNCTION_TARGET", "accb-dev-gcp-ocr-handler"),
                "dedupe_key": dedupe_key,
                "decision": decision,
                "latency_ms": int((time.time() - started) * 1000),
            }
        )
    )

File: examples/canonical-gcp-functions/gcs-trigger-ocr-to-firestore/typescript/src/handler.js
async function handleGcsObject(event, clients) {
  const started = Date.now();
  const activeClients = clients ?? gcpClients();
  const generation = String(event.data.generation ?? event.data.metageneration ?? "0");
  const eventId = event.id ?? `${event.data.bucket}/${event.data.name}/${generation}`;
  const dedupeKey = `${event.data.bucket}:${event.data.name}:${generation}`;
  const claimed = await activeClients.claimGeneration(dedupeKey, { ...event.data, generation });
  if (!claimed) {
    log(started, eventId, dedupeKey, "DROP", "duplicate object generation");
    return { ok: true, duplicate: true, decision: "DROP" };
  }
  const text = await activeClients.runOcr(event.data.bucket, event.data.name, generation);
  const payload = { ...event.data, generation, text, status: "OCR_COMPLETE" };
  await activeClients.writeMetadata(dedupeKey, payload);
  await activeClients.publishDownstream(payload);
  log(started, eventId, dedupeKey, "ALLOW", "ocr metadata written");
  return { ok: true, duplicate: false, decision: "ALLOW", text };
}
function gcpClients() { throw new Error("GCP clients are bound by deployment; tests inject clients."); }
function log(started, correlationId, dedupeKey, decision, msg) {
  console.log(JSON.stringify({ timestamp: new Date().toISOString(), severity: "INFO", msg, trace_id: process.env.TRACEPARENT ?? "local-trace", request_id: correlationId, correlation_id: correlationId, provider: "gcp", runtime_tier: "function", function_name: process.env.FUNCTION_TARGET ?? "accb-dev-gcp-gcs-ocr-typescript", dedupe_key: dedupeKey, decision, latency_ms: Date.now() - started }));
}
module.exports = { handleGcsObject };
